cmd/dashboard-redirector/main.go

```diff
@@ -0,0 +1,109 @@
+//
+// Copyright (c) 2018 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"net/http"
+	"strings"
+
+	crd "github.com/openshift/ansible-service-broker/pkg/dao/crd"
+
+	"github.com/sirupsen/logrus"
+)
+
+var options struct {
+	BrokerNamespace string
+	Port            int
+}
+
+func init() {
+	flag.IntVar(&options.Port, "port", 1337, "port that the dashboard-redirector should listen on")
+	flag.StringVar(&options.BrokerNamespace, "namespace", "ansible-service-broker", "namespace that the broker resides in")
+	flag.Parse()
+}
+
+var crdDao *crd.Dao
+
+func main() {
+	var err error
+
+	crdDao, err = crd.NewDao(options.BrokerNamespace)
+	if err != nil {
+		panic(fmt.Sprintf("Unable to create crd client - %v", err))
+	}
+
+	logrus.Info("Trying to load batch specs as a sanity check...")
+	specs, err := crdDao.BatchGetSpecs("")
+	if err != nil {
+		panic(fmt.Sprintf("Sanity check failed! -> %v", err))
+	} else {
+		logrus.Info("Sanity check passed! Loaded specs: %v", specs)
+	}
+
+	http.HandleFunc("/", redirect)
+	portStr := fmt.Sprintf(":%d", options.Port)
+
+	logrus.Infof("Dashboard redirector listening on port [%s]", portStr)
+	err = http.ListenAndServe(portStr, nil)
+	if err != nil {
+		logrus.Fatal("ListenAndServe: ", err)
+	}
+}
+
+func redirect(w http.ResponseWriter, r *http.Request) {
+	var errMsg string
+	logrus.Info("Checking for form")
+
+	id := r.FormValue("id")
+	if id == "" {
+		errMsg := "Did not find service instance id as a query param!"
+		logrus.Error(errMsg)
+		http.Error(w, errMsg, http.StatusBadRequest)
+		return
+	}
+
+	logrus.Infof("Got request for service instance %s, looking up dashboard_url", id)
+
+	si, err := crdDao.GetServiceInstance(id)
+	if err != nil {
+		errMsg = fmt.Sprintf("Something went wrong trying to load service instance [%s] -> %s", id, err)
+		logrus.Errorf(errMsg, id, err.Error())
+		http.Error(w, errMsg, http.StatusInternalServerError)
+		return
+	}
+
+	logrus.Info("Successfully loaded SI: %+v", si)
+
+	if si.DashboardURL == "" {
+		errMsg = fmt.Sprintf("No DashboardURL set for requested instance! %v", id)
+		logrus.Infof("%s, returning 404", errMsg)
+		http.Error(w, errMsg, http.StatusNotFound)
+		return
+	}
+
+	logrus.Infof("DashboardURL found: %s, 301 redirecting", si.DashboardURL)
+	var redirectURL string
+	if !strings.HasPrefix("http", si.DashboardURL) {
+		redirectURL = fmt.Sprintf("http://%s", si.DashboardURL)
```

Comment:
What's the use case behind this? Is it possible to just force whatever external entity is supplying the URL to supply a URL with a valid scheme?

Aside from the risk of assuming non-tls http, it's possible a service would have a management dashboard available through other non-http means, such as ssh.

If there is a use case that requires making this assumption, it would be helpful to add an in-line comment about it.

Reply:
> What's the use case behind this? Is it possible to just force whatever external entity is supplying the URL to supply a URL with a valid scheme?

When an APB passes back something without a scheme; this is more friendly and allows that case to pass and assume http rather than explicitly fail.

> Aside from the risk of assuming non-tls http

It seems to me the worst case scenario here is that http is assumed when not supported by the service. Many will force a redirect to https.

> it's possible a service would have a management dashboard available through other non-http means, such as ssh.

Interesting point.

I'm anticipating there will be APB authors that unintentionally do not provide a scheme and the intention is to help them with a reasonable assumption. Alternatively, we can lock this down, warn, and default to not providing the dashboard url if the scheme is missing. As I'm thinking about it, the latter might not be a bad idea since the majority of authors will probably get that service URL dynamically from some API and store it in a var to later provide to our module, greater increasing the likelyhood of having the scheme handled by the remote system.